### ISSUE TITLE
Fluidsynth doesn't respond to either `SIGTERM` nor `SIGINT` in some scenarios

### ISSUE THREAD

**Issue by itaranto** [FluidSynth/fluidsynth#1071]
### FluidSynth version

```
FluidSynth executable version 2.2.5
Sample type=double
```

### Describe the bug

Fluidsynth doesn't respond to either `SIGTERM` nor `SIGINT` in some scenarios.

The only way to terminate the `fluidsynth` process is by using `SIGKILL`.

### Expected behavior

Fluidsynth should exit gracefully upon receiving `SIGTERM` or `SIGINT`.

### Steps to reproduce

1. Run Fluidsynth like this:

```sh
fluidsynth \
    --no-shell \
    --server \
    --audio-driver=pulseaudio \
    --midi-driver=alsa_seq \
    --sample-rate 48000 \
    /usr/share/soundfonts/FluidR3_GM.sf2
```

2. Press Ctrl+C

Or:

1. Run Fluidsynth like this:

```sh
fluidsynth \
    --no-shell \
    --server \
    --audio-driver=pulseaudio \
    --midi-driver=alsa_seq \
    --sample-rate 48000 \
    /usr/share/soundfonts/FluidR3_GM.sf2 &
```
2. Find Fluidsynth's PID

3. Run `kill <PID>`

**Comment by derselbst** [comment]
Sry, I cannot reproduce any of your examples. Can you pls. test with a different audio driver like ALSA?

**Comment by itaranto** [comment]
> Sry, I cannot reproduce any of your examples. Can you pls. test with a different audio driver like ALSA?

Sorry I made a copy-paste error, I've tried with ALSA too:

Like this:

```sh
fluidsynth \
    --no-shell \
    --server \
    --audio-driver=alsa \
    --midi-driver=alsa_seq \
    --sample-rate 48000 \
    /usr/share/soundfonts/FluidR3_GM.sf2
```

I'm using Pipewire if that helps.

**Comment by derselbst** [comment]
Sry, I don't understand. Do you want to say that you originally used pipewire? Meaning, that you're using recent master and not 2.2.5? And no matter if you use ALSA, pulse or pipewire, it doesn't respond to SIGTERM and SIGINT?

**Comment by derselbst** [comment]
Ok. I have no experience with pipewire, so I don't know what's happening under the hood.

What happens if you omit the `--server` flag?

**Comment by mawe42** [comment]
This sounds very much like the bug I experienced:
https://lists.nongnu.org/archive/html/fluid-dev/2022-03/msg00005.html

**Comment by mawe42** [comment]
@itaranto Can you please check if FluidSynth exists properly if you start it as you do (including `--server`), then `SIGINT` or `SIGTERM` it and do a `telnet localhost 9800`. If it is the same bug I experienced, then FluidSynth should exit immediately after you connect to it via telnet.

**Comment by itaranto** [comment]
> Ok. I have no experience with pipewire, so I don't know what's happening under the hood.
> 
> What happens if you omit the `--server` flag?

If I remove the `--server` flag I have to also remove the `--no-shell` flag or no process will be run.

So, if run:

```sh
fluidsynth \
    --audio-driver=alsa \
    --midi-driver=alsa_seq \
    --sample-rate 48000 \
    /usr/share/soundfonts/FluidR3_GM.sf2
```

I see:
```
ALSA lib pcm_dsnoop.c:601:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_a52.c:1001:(_snd_pcm_a52_open) a52 is only for playback
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
FluidSynth runtime version 2.2.5
Copyright (C) 2000-2022 Peter Hanappe and others.
Distributed under the LGPL license.
SoundFont(R) is a registered trademark of Creative Technology Ltd.

fluidsynth: warning: Failed to set thread to high priority
fluidsynth: Using PulseAudio driver
Type 'help' for help topics.

(ins)> fluidsynth: warning: Failed to set thread to high priority
```

The process still cannot be stoped without a `SIGKILL`, after sending `SIGKILL` I see:

```
fish: Job 1, 'fluidsynth \…' terminated by signal --audio-dr… (SIGKILL)
fish: Job Forced quit, '' terminated by signal  ()
```

*but*

I realized that sending Ctrl+D makes fluidsynth to stop:

```
fluidsynth: panic: An error occurred while reading from stdin.
```

**Comment by itaranto** [comment]
> @itaranto Can you please check if FluidSynth exits properly if you start it as you do (including `--server`), then `SIGINT` or `SIGTERM` it and then do a `telnet localhost 9800`. If it is the same bug I experienced, then FluidSynth should exit immediately after you connect to it via telnet.

I don't have `telnet` installed but I do have `openbsd-netcat`, After running FluidSynth, in another console I ran `nc localhost 9800` and typed `help`, this is what I see:

```

*** Help topics:***
help all (prints all topics)
help general
help event
help polymono
help reverb
help chorus
help tuning
help settings
help ladspa
help router
help player
```

But FluidSynth keeps running...

**Comment by mawe42** [comment]
Just to be sure, because in your reply you didn't mention any `kill` command: please try the following:
1. Start FluidSynth with `--server` and all your other standard options
2. `kill` FluidSynth with `SIGINT` or `SIGTERM`
3. Connect to port 9800 with telnet or netcat

If the problem is the same as the one I experienced, then FluidSynth should keep running after the `kill` but should exit immediately after connecting to port 9800.

**Comment by itaranto** [comment]
> Just to be sure, because in your reply you didn't mention any `kill` command: please try the following:
> 
>     1. Start FluidSynth with `--server` and all your other standard options
> 
>     2. `kill` FluidSynth with `SIGINT` or `SIGTERM`
> 
>     3. Connect to port 9800 with telnet or netcat
> 
> 
> If the problem is the same as the one I experienced, then FluidSynth should keep running after the `kill` but should exit immediately after connecting to port 9800.

Nope, I ran `fluidsynth` the same as above, attemped to kill it with `kill` (uses `SIGTERM` by default), and then `nc localhost 9800`, but nothing happened.

**Comment by derselbst** [comment]
As we can see, even though you specified `--audio-driver=alsa`, pluseaudio was used as audio driver. This is strange. IMO, you should ensure to test a different audio driver. Watch out for fluidsynth saying `fluidsynth: Using *** driver`.

If the problem does not occur when using a driver different than pulseaudio, it's likely a pulseaudio issue and we cannot do anything about it. You could also try installing a different version of pulseaudio and see what happens.

**Comment by itaranto** [comment]
> As we can see, even though you specified `--audio-driver=alsa`, pluseaudio was used as audio driver. This is strange. IMO, you should ensure to test a different audio driver. Watch out for fluidsynth saying `fluidsynth: Using *** driver`.
> 
> If the problem does not occur when using a driver different than pulseaudio, it's likely a pulseaudio issue and we cannot do anything about it. You could also try installing a different version of pulseaudio and see what happens.

I'm sorry for being so sloppy,  I messed up the console paste *again*, thanks to `fish` autocomplete :(

This is the real output when running with `--audio-driver=alsa`:

```
ALSA lib pcm_dsnoop.c:601:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_a52.c:1001:(_snd_pcm_a52_open) a52 is only for playback
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
FluidSynth runtime version 2.2.6
Copyright (C) 2000-2022 Peter Hanappe and others.
Distributed under the LGPL license.
SoundFont(R) is a registered trademark of Creative Technology Ltd.

fluidsynth: warning: Failed to set thread to high priority
fluidsynth: warning: Failed to set thread to high priority
```

As you can see, there's no `Using PulseAudio driver`.

**Comment by itaranto** [comment]
Anyway, I looked at the source code and I don't see any mention for `SIGINT`, `SIGTERM` nor the `signal` C function.

Are you guys sure this is actually implemented in FluidSynth?

**Comment by derselbst** [comment]
> As you can see, there's no Using PulseAudio driver message.

Sry, I forgot to mention that you need to either pass `--verbose` or compile fluidsynth in debug mode to see it.

Yet, there're way too many unrelated logs coming out. All those jack related logs like "jack server is not running or cannot be started" should not be seen when using ALSA. I don't know what going on on your system.

> Anyway, I looked at the source code and I don't see any mention of SIGINT, SIGTERM nor the signal or sigaction C functions. Are you guys sure this is actually implemented in FluidSynth?

Fluidsynth is not intercepting signals. This is why believe it's coming from elsewhere.

**Comment by itaranto** [comment]
> Sry, I forgot to mention that you need to either pass `--verbose` or compile fluidsynth in debug mode to see it.

OK, with the `--verbose` and  `--audio-driver=alsa` flags I now see: `fluidsynth: debug: Using 'alsa' audio driver`

> Yet, there're way too many unrelated logs coming out. All those jack related logs like "jack server is not running or cannot be started" should not be seen when using ALSA. I don't know what going on on your system

Me neither, but I think this has nothing to do with system signals.

> Fluidsynth is not intercepting signals. This is why believe it's coming from elsewhere.

I'm not sure if I'm following you, but FluidSynth not handling those signals it's the whole point of this bug.

**Comment by derselbst** [comment]
> I'm not sure if I'm following you, but FluidSynth not handling those signals it's the whole point of this bug
Just want to make clear that FluidSynth works with my current use case (MIDI playback), it's just that it doesn't handle standard Unix signals.

If a program doesn't implement signal handlers (like fluidsynth), the [default action](https://man7.org/linux/man-pages/man7/signal.7.html) takes place when a signal is received. On POSIX-compliant systems, this default action is to terminate the program for both SIGTERM and SIGINT.

This is why I believe something else is responsible for intercepting and ignoring those signals. Or causing a different kind of deadlock.

So, are you on a POSIX-compliant operating system? If yes, pls. start fluidsynth with gdb, set up breakpoints for `signal` and `sigaction` and post the backtrace when you hit it.

**Comment by itaranto** [comment]
> If a program doesn't implement signal handlers (like fluidsynth), the [default action](https://man7.org/linux/man-pages/man7/signal.7.html) takes place when a signal is received. On POSIX-compliant systems, this default action is to terminate the program for both SIGTERM and SIGINT.

> This is why I believe something else is responsible for intercepting and ignoring those signals. Or causing a different kind of deadlock.

Yes, that makes sense.

> So, are you on a POSIX-compliant operating system? If yes, pls. start fluidsynth with gdb, set up breakpoints for `signal` and `sigaction` and post the backtrace when you hit it.

I'm on Arch Linux, there's a `fluidsynth-git` AUR package that compiles FluidSynth from master, I'll try either that or compiling by hand.

**Comment by derselbst** [comment]
@itaranto Any news?

**Comment by itaranto** [comment]
> @itaranto Any news?

Sorry I've been busy lately I'll test it on the weekend.

**Comment by itaranto** [comment]
OK, so I compiled from `master` (`a0ca05e16d844cf483a89105bfd3296699a252f3`) and executed the binary from `build/src` in the same way as a above and the issue persists.

**Comment by derselbst** [comment]
Have you set breakpoints for signal and sigaction to see if they are called by something?

**Comment by mawe42** [comment]
> Yet, there're way too many unrelated logs coming out. All those jack related logs like "jack server is not running or cannot be started" should not be seen when using ALSA. I don't know what going on on your system.

All those error messages come from PortAudio. Pa_Initialize potentially triggers lots of error messages to stdout. It seems to be a known problem with PortAudio. And as Pa_Initialize is called in `fluid_portaudio_driver_settings()`, simply having it compiled into the fluidsynth binary means you will see those messages. But this is unrelated to the reported problem, I think.

I could recreate the problem on my system by compiling in the SDL2 audio driver. It seems like `SDL_Init()` installs it's own signal handlers for SIGINT and SIGTERM, turning them into a "Quit" event on the SDL event loop. That's why those signals never reach Fluidsynth and have no effect.

Giving SDL a "hint" that we want it to leave signals alone seems to fix the problem and causes Fluidsynth to actually exit on SIGTERM and SIGINT:
```
#if SDL2_SUPPORT
    // Tell SDL that it shouldn't try to handle signals
    SDL_SetHint(SDL_HINT_NO_SIGNAL_HANDLERS, "1");

    if(SDL_Init(SDL_INIT_AUDIO) != 0)
    {
        fprintf(stderr, "Warning: Unable to initialize SDL2 Audio: %s", SDL_GetError());
    }
    else
    {
        atexit(SDL_Quit);
    }
#endif
```

**Comment by derselbst** [comment]
> I could recreate the problem on my system by compiling in the SDL2 audio driver.

Interesting. I also have compiled with SDL2 support, but it works for me, no matter what audio driver I use. I'm using SDL 2.0.8, which version do you have?

@itaranto Could you pls. apply the change Markus has suggested to `fluidsynth.c`, recompile and let us know the result?

**Comment by mawe42** [comment]
> I also have compiled with SDL2 support, but it works for me, no matter what audio driver I use. I'm using SDL 2.0.8, which version do you have?

I use SDL 2.0.20. Maybe they changed the default behaviour sometime between 2.0.8 and 2.0.20?

**Comment by derselbst** [comment]
Aha, now I can reproduce it, thanks! Just committed the fix for fluidsynth 2.2.7 (4ee4b0403e9f461e8bff18a6a0e45f3731fb40cf). Waiting for @itaranto to give the final confirmation.

**Comment by itaranto** [comment]
> Aha, now I can reproduce it, thanks! Just committed the fix for fluidsynth 2.2.7 ([4ee4b04](https://github.com/FluidSynth/fluidsynth/commit/4ee4b0403e9f461e8bff18a6a0e45f3731fb40cf)). Waiting for @itaranto to give the final confirmation.

I applied that fix and it works both `SIGTERM` and `SIGINT` by using either `pkill` or `kill`.

The strange thing is, it doesn't seem to respond to Ctrl + C while using the `fish` shell, it works on bash though.

Although the fix is good enough for the moment, I need to investigate what's going on with `fish`.

**Comment by mawe42** [comment]
> It also works with Ctrl+C, so I think this issue can be closed.

Great! And it seems this also fixes the problem I experienced a few weeks ago, mentioned in https://github.com/FluidSynth/fluidsynth/issues/1071#issuecomment-1074345582

A slightly different effect, but it seems the SDL signal handlers prevented the `accept()` from interrupting on signal.

**Comment by derselbst** [comment]
Merged and will be available in 2.2.7
